cdc/cdc.cc

```diff
@@ -0,0 +1,221 @@
+    return when_all(remove_log(ctx, ks_name, table_name),
+                    remove_desc(ctx, ks_name, table_name)).discard_result();
+}
+
+static future<> setup_log(db_context ctx, const schema& s) {
+    schema_builder b(s.ks_name(), log_name(s.cf_name()));
+    b.set_default_time_to_live(gc_clock::duration{86400}); // 24h
+    b.set_comment(sprint("CDC log for %s.%s", s.ks_name(), s.cf_name()));
+    b.with_column("stream_id", uuid_type, column_kind::partition_key);
+    b.with_column("time", timeuuid_type, column_kind::clustering_key);
+    b.with_column("batch_seq_no", int32_type, column_kind::clustering_key);
+    b.with_column("operation", int32_type);
+    b.with_column("ttl", long_type);
+    auto add_columns = [&] (const auto& columns) {
+        for (const auto& column : columns) {
+            b.with_column("_" + column.name(), column.type);
+        }
+    };
+    add_columns(s.partition_key_columns());
+    add_columns(s.clustering_key_columns());
+    add_columns(s.static_columns());
+    add_columns(s.regular_columns());
+    return ctx._migration_manager.announce_new_column_family(b.build(), false);
+}
+
+static future<> setup_desc(db_context ctx, const schema& s) {
+    schema_builder b(s.ks_name(), desc_name(s.cf_name()));
+    b.set_comment(sprint("CDC description for %s.%s", s.ks_name(), s.cf_name()));
+    b.with_column("node_ip", inet_addr_type, column_kind::partition_key);
+    b.with_column("shard_id", int32_type, column_kind::partition_key);
+    b.with_column("created_at", timestamp_type, column_kind::clustering_key);
+    b.with_column("stream_id", uuid_type);
+    return ctx._migration_manager.announce_new_column_family(b.build(), false);
+}
+
+template <typename Func>
+static void for_each_endpoint(db_context ctx, Func&& f) {
+    for (const auto& dc : ctx._token_metadata.get_topology().get_datacenter_endpoints()) {
+        for (const auto& endpoint : dc.second) {
+            f(endpoint);
+        }
+    }
+}
+
+template <typename Func>
+static void
+for_each_shard(db_context ctx, const gms::inet_address& endpoint, Func&& f) {
+    const unsigned int shard_count = ctx._snitch->get_shard_count(endpoint);
+    for (unsigned int shard_id = 0; shard_id < shard_count; ++shard_id) {
+        f(shard_id, shard_count);
+    }
+}
+
+// This function assumes setup_desc was called on |s| before the call to this
+// function.
+static future<> populate_desc(db_context ctx, const schema& s) {
+    auto& db = ctx._proxy.get_db().local();
+    auto desc_schema =
+        db.find_schema(s.ks_name(), desc_name(s.cf_name()));
+    auto log_schema =
+        db.find_schema(s.ks_name(), log_name(s.cf_name()));
+    // The assumption here is that all nodes in the cluster are using the
+    // same murmur3_partitioner_ignore_msb_bits. If it's not true, we will
+    // have to gossip it around together with shard_count.
+    auto ignore_msb_bits =
+        db.get_config().murmur3_partitioner_ignore_msb_bits();
+
```

Comment:
should use global_partitioner().sharding_ignore_msb() since that's what the system uses (it derives it from config, but better get it from the source of truth)

Reply:
done.